Convert to IIncrementalGenerator

davhdavh opened the pull request:
Tried to run it on a project with a couple of thousand resx files and ran into a couple of issues.
1. Dont generate invalid namespace for root level resx files
2. Dont generate invalid classnames when resx files have multiple extensions
3. Give compiler warning on duplicate members
4. Give compiler warning when member has same name as class
5. Convert to IIncrementalGenerator + instant intellisense when changing file

davhdavh committed:
1. Dont generate invalid namespace for root level resx files
2. Dont generate invalid classnames when resx files have multiple extensions
3. Give compiler warning on duplicate members
4. Give compiler warning when member has same name as class

davhdavh committed:
Convert to IIncrementalGenerator, intellisense is instant when adding/changing files now, even in huge project

ycanardeau commented on VocaDb.ResXFileCodeGenerator/FileOptions.cs:
```diff
@@ -0,0 +1,59 @@
+
+namespace VocaDb.ResXFileCodeGenerator;
+
+internal record FileOptions
```
Please apply the `sealed` modifier.

davhdavh replied:
old java dev? It doesn't matter in C# since the default is non-virtual. I have applied it nonetheless.

ycanardeau replied:
It's true that I used to use Java many years ago, but I tend to mark classes as sealed as much as possible unless they really need to be inherited (I think classes are sealed by default in newer languages). It's relatively easy to remove the sealed modifier, but hard to add it afterwards (if they have some derived classes already). But it's not a big problem anyway as long as the class is marked as internal.

ycanardeau commented on VocaDb.ResXFileCodeGenerator.Tests/GeneratorTests.cs:
```diff
@@ -108,7 +108,7 @@ namespace Resources
     }}
 }}";
-		using var resxStream = new MemoryStream(Encoding.UTF8.GetBytes(Text));
+		using var resxStream = new StringReader(Text);
```
Can you please explain why you replaced `MemoryStream` with `StringReader`? I guess because of performance reasons?

davhdavh replied:
Because the built-in method of reading the file uses SourceText object, and there was no particular reason to convert to utf8 bytes just to convert it back a couple of lines later.

ycanardeau commented on VocaDb.ResXFileCodeGenerator/FileOptions.cs:
```diff
@@ -0,0 +1,59 @@
+		Valid = globalOptions.Valid;
+	}
+
+	public AdditionalText File { get; }
+
+	public string FilePath { get; }
+	public bool StaticClass { get; }
+	public bool NullForgivingOperators { get; }
+	public bool PublicClass { get; }
+	public string ClassName { get; }
+	public string? CustomToolNamespace { get; }
+	public string LocalNamespace { get; }
+	public bool Valid { get; }
```
Can you please put these properties before the constructor?

davhdavh replied:
done

ycanardeau commented on VocaDb.ResXFileCodeGenerator/GlobalOptions.cs:
```diff
@@ -0,0 +1,51 @@
+		return new(provider.GlobalOptions);
+	}
+
+	public string RootNamespace { get; }
+
+	public string ProjectFullPath { get; }
+
+	public bool Valid { get; }
+
+	public bool StaticClass { get; }
+
+	public bool NullForgivingOperators { get; }
+
+	public bool PublicClassGlobal { get; }
```
Can you please put these properties before the constructor?

davhdavh replied:
done

ycanardeau commented on VocaDb.ResXFileCodeGenerator/VocaDb.ResXFileCodeGenerator.csproj:
```diff
@@ -15,7 +15,7 @@
     <NoWarn>$(NoWarn);NU5128</NoWarn>
     <DevelopmentDependency>true</DevelopmentDependency>
-    <Version>2.0.2</Version>
+    <Version>3.0.0</Version>
```
If there are no breaking changes in this PR, I think bumping the version number to either `2.0.3` or `2.1.0` makes more sense.

davhdavh replied:
True, there are however substantial speed increases and the generation is based on what the editor says the file contains rather than the file content. I personally still think it should be a 3.0.
Also, I am adding the partial class generation we discussed in issue #5, that will be included in next commit.

ycanardeau commented on VocaDb.ResXFileCodeGenerator/GlobalOptions.cs:
```diff
@@ -0,0 +1,51 @@
+		                staticClassSwitch is { Length: > 0 } &&
+		                staticClassSwitch.Equals("false", StringComparison.OrdinalIgnoreCase));
+
+		Valid = true;
```
Why is the `Valid` property needed? I can see this field is set only once here.

ycanardeau commented on VocaDb.ResXFileCodeGenerator/SourceGenerator.cs:
```diff
@@ -1,146 +1,41 @@
+		var inputs = monitor
+			.Combine(globalOptions)
+			.Combine(context.AnalyzerConfigOptionsProvider)
+			.Select((x,t)=>FileOptions.Select(x.Left.Left, x.Right, x.Left.Right, t))
+			.Where(x=>x.Valid);
```
Style nit: Please place spaces before and after the `=>` operator.

davhdavh replied:
done

ycanardeau commented on VocaDb.ResXFileCodeGenerator.Tests/GithubIssues/Issue3/GeneratorTests.cs:
```diff
@@ -116,4 +117,141 @@ public static class CommonMessages
+				PublicClass: true,
+				NullForgivingOperators: false,
+				StaticClass: true,
+				ReportError: e=>errs.Add(e),
```
Style nit: Please place spaces before and after the `=>` operator.

davhdavh replied:
done

ycanardeau commented on VocaDb.ResXFileCodeGenerator.Tests/GithubIssues/Issue3/GeneratorTests.cs:
```diff
@@ -116,4 +117,141 @@ public static class CommonMessages
+				PublicClass: true,
+				NullForgivingOperators: false,
+				StaticClass: true,
+				ReportError: e=>errs.Add(e),
```
Style nit: Please place spaces before and after the `=>` operator.

davhdavh replied:
done

ycanardeau commented on VocaDb.ResXFileCodeGenerator/SourceGenerator.cs:
```diff
@@ -1,146 +1,41 @@
-			context.AddSource($"{options.LocalNamespace}.{options.ClassName}.g.cs", source);
-		}
+			var filecontent = file.File.GetText(ctx.CancellationToken);
+			if (filecontent == null) return;
```
Please replace `== null` with `is null`.

davhdavh replied:
done

ycanardeau commented on VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs:
```diff
@@ -3,15 +3,33 @@
 public sealed class StringBuilderGenerator : IGenerator
 {
 	private static readonly Regex ValidMemberNamePattern = new(@"^[\p{L}\p{Nl}_][\p{Cf}\p{L}\p{Mc}\p{Mn}\p{Nd}\p{Nl}\p{Pc}]*$", RegexOptions.Compiled | RegexOptions.CultureInvariant);
 	private static readonly Regex InvalidMemberNameSymbols = new(@"[^\p{Cf}\p{L}\p{Mc}\p{Mn}\p{Nd}\p{Nl}\p{Pc}]", RegexOptions.Compiled | RegexOptions.CultureInvariant);
-	public string Generate(Stream resxStream, GeneratorOptions options)
+
+	private static readonly DiagnosticDescriptor DuplicateError = new(id: "VocaDbResXFileCodeGenerator001",
+		title: "Duplicate member",
+		messageFormat: "Ignored added member '{0}'",
+		category: "ResXFileCodeGenerator",
+		DiagnosticSeverity.Warning,
+		isEnabledByDefault: true);
+
+	private static readonly DiagnosticDescriptor MemberSameAsClassError = new(id: "VocaDbResXFileCodeGenerator002",
+		title: "Member same name as class",
+		messageFormat: "Ignored member '{0}' has same name as class",
+		category: "ResXFileCodeGenerator",
+		DiagnosticSeverity.Warning,
+		isEnabledByDefault: true);
```
Can you please prefix private static fields with `s_`?

davhdavh replied:
done

ycanardeau commented on VocaDb.ResXFileCodeGenerator/GlobalOptions.cs:
```diff
@@ -0,0 +1,51 @@
+
+namespace VocaDb.ResXFileCodeGenerator;
+
+internal record GlobalOptions
```
Please apply the `sealed` modifier.

davhdavh committed:
1. #5 add ability to generate inner classes
2. Apply review updates

davhdavh commented:
I didn't actually test this yet, so just review it, don't merge it.

ycanardeau commented on VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs:
```diff
@@ -61,39 +126,66 @@ public string Generate(Stream resxStream, GeneratorOptions options)
 
-			builder.AppendLine("        /// <summary>");
+			static Location GetMemberLocation(FileOptions fileOptions, IXmlLineInfo line, string memberName) =>
+				Location.Create(fileOptions.FilePath, new(),
+					new(new(line.LineNumber - 1, line.LinePosition - 1),
+						new(line.LineNumber - 1, line.LinePosition - 1 + memberName.Length)));
```
I would use named arguments for better readability here, but I can do it after merging this PR.

ycanardeau commented on VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs:
```diff
@@ -35,17 +63,54 @@ public string Generate(Stream resxStream, GeneratorOptions options)
+				if (options.StaticClass || options.StaticMembers)
+				{
+					reportError?.Invoke(Diagnostic.Create(s_memberWithStaticError,
+						Location.Create(options.FilePath!, new(), new())));
```
I think the null-forgiving operator is not needed, since `FilePath` is not null here.

davhdavh replied:
Done, I cleaned it up below, but forgot here

ycanardeau commented on VocaDb.ResXFileCodeGenerator/FileOptions.cs:
```diff
@@ -0,0 +1,114 @@
+	public InnerClassVisibility InnerClassVisibility { get; init; }
+	public bool PartialClass { get; init; }
+	public bool StaticMembers { get; init; } = true;
+	public AdditionalText File { get; init; }
```
Isn't it possible to make the `File` property nullable rather than assigning `null!` to this property in the constructor?
I think `file.File.GetText(ctx.CancellationToken)` can be replaced with `file.File?.GetText(ctx.CancellationToken)` in the `SourceGenerator` class.

davhdavh replied:
Ofcourse it is. However, it is only ever null in testing, thus my opinion is that it should self document that this cannot ever be null in real code.

ycanardeau commented on VocaDb.ResXFileCodeGenerator/FileOptions.cs:
```diff
@@ -0,0 +1,114 @@
+
+	public FileOptions()
+	{
+		LocalNamespace = "";
+		CustomToolNamespace = "";
+		ClassName = "";
+		InnerClassInstanceName = "";
+		InnerClassName = "";
+		File = null!;
+		FilePath = "";
```
I'd replace these with `string.Empty`, but not mandatory.

davhdavh replied:
done

ycanardeau commented on README.md:
```diff
@@ -115,7 +127,195 @@ or globally
+
+### Inner Class Visibility (per file or globally)
+
+Per default inner classes are not generated, unless this setting is one of the following:
```
I'm not a native English speaker, but I guess "by default" would be better?

davhdavh replied:
They are interchangeable, but sure, fixed

ycanardeau commented on README.md:
```diff
@@ -115,7 +127,195 @@ or globally
+
+### Inner Class name (per file or globally)
+
+Per default the inner class is named "Resources", which can be overriden with this setting:
```
here

ycanardeau commented on README.md:
```diff
@@ -115,7 +127,195 @@ or globally
+
+### Inner Class instance name (per file or globally)
+
+Per default no instance is available of the class, but that can be made available if this setting is given.
```
and here

ycanardeau commented on README.md:
```diff
@@ -115,7 +127,195 @@ or globally
+<ItemGroup>
+	<EmbeddedResource Update="**/*.resx">
+		<DependentUpon>$([System.String]::Copy('%(FileName).cs'))</DependentUpon>
+		<InnerClassName>MyResources</InnerClassName>
+		<InnerClassVisibility>private</InnerClassVisibility>
+		<InnerClassInstanceName>EveryoneLikeMyNaming</InnerClassInstanceName>
+	    <StaticMembers>false</StaticMembers>
```
You need to add these new options to [VocaDb.ResXFileCodeGenerator.props](https://github.com/VocaDB/ResXFileCodeGenerator/blob/main/VocaDb.ResXFileCodeGenerator/build/VocaDb.ResXFileCodeGenerator.props), so that these properties are [visible to the compiler](https://github.com/dotnet/roslyn/blob/main/docs/features/source-generators.cookbook.md#consume-msbuild-properties-and-metadata).

davhdavh replied:
Ahh cool, didn't even know that.

ycanardeau commented on README.md:
````diff
@@ -115,7 +127,195 @@ or globally
+
+```xml
+<PropertyGroup>
+  <ResXFileCodeGenerator_InnerClassName>MyResources</ResXFileCodeGenerator_InnerClassName>
+  <ResXFileCodeGenerator_InnerClassVisibility>private</ResXFileCodeGenerator_InnerClassVisibility>
+  <ResXFileCodeGenerator_InnerClassInstanceName>EveryoneLikeMyNaming</InnerClassInstanceName>
+  <ResXFileCodeGenerator_StaticMembers>false</ResXFileCodeGenerator_StaticMembers>
````
these as well

ycanardeau commented on VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs:
```diff
@@ -35,17 +63,54 @@ public string Generate(Stream resxStream, GeneratorOptions options)
+				builder.Append(options.ClassName);
+				builder.Append(" ");
+				builder.Append(options.InnerClassInstanceName);
+				builder.Append(Constants.CultureInfoVariable);
```
Is it intentional that `Constants.CultureInfoVariable` is appended here?

davhdavh replied:
Nope, good catch, and the unittest was wrong also.

ycanardeau commented on VocaDb.ResXFileCodeGenerator/FileOptions.cs:
```diff
@@ -0,0 +1,114 @@
+		StaticMembers =
+			options.TryGetValue("build_metadata.EmbeddedResource.StaticMembers", out var staticMembersSwitch) &&
+			staticMembersSwitch is { Length: > 0 }
+				? staticMembersSwitch.Equals("false", StringComparison.OrdinalIgnoreCase)
```
This line needs to be negated just like above, otherwise it generates non-static members when I set this option to `true`. It would be nice if you could write unit tests for options classes.

davhdavh replied:
Indeed, fixed

ycanardeau commented on VocaDb.ResXFileCodeGenerator/FileOptions.cs:
```diff
@@ -0,0 +1,114 @@
+
+public sealed record FileOptions
+{
+	public string InnerClassInstanceName { get; init; }
+	public string InnerClassName { get; init; }
+	public InnerClassVisibility InnerClassVisibility { get; init; }
+	public bool PartialClass { get; init; }
+	public bool StaticMembers { get; init; } = true;
```
Can you please add comments for these options as the number of options has been increasing? Copying and pasting the description from readme should be enough.

ycanardeau commented on VocaDb.ResXFileCodeGenerator/GlobalOptions.cs:
```diff
@@ -0,0 +1,89 @@
+
+internal sealed class GlobalOptions
+{
+	public string InnerClassInstanceName { get; }
+	public bool StaticMembers { get; }
+	public string InnerClassName { get; }
+	public InnerClassVisibility InnerClassVisibility { get; }
```
For these too.

davhdavh committed:
Code review and add unit test for settings

davhdavh committed:
Fix bug when using inner classes and have members with same name as outer class name
Fix non-bug extra line added when ignoring duplicate member or member with same name as class

ycanardeau commented on VocaDb.ResXFileCodeGenerator.Tests/GeneratorTests.cs:
```diff
@@ -139,198 +145,248 @@ namespace Resources
-        /// </summary>
-        public static string CreateDateDescending => ResourceManager.GetString(nameof(CreateDateDescending), CultureInfo)!;
+        {(publicClass ? "public" : "internal")}{(partial ? " partial" : "")}{(staticClass ? " static" : "")} class {innerClassName}
+        {{{(string.IsNullOrEmpty(innerClassInstanceName) ? "" : $"\n            public {innerClassInstanceName} {{ get; }} = new();\n")}
```
The type name is missing, which causes failing tests. By the way, is it the intended behavior that this property is generated for the "inner" class? Should it be generated for the container class instead as you mentioned in #5? I know you're still editing...

davhdavh replied:
Indeed, I made a total mess of that

ycanardeau commented on VocaDb.ResXFileCodeGenerator/FileOptions.cs:
```diff
@@ -0,0 +1,118 @@
+		Valid = globalOptions.Valid;
+	}
+
+	public static FileOptions Select(AdditionalText file, AnalyzerConfigOptionsProvider options,
```
For your information, you can make it internal by using `InternalsVisibleTo`, but I think it's okay to be public for now.

davhdavh replied:
I know, doesn't matter in this case since its not a library anyone is ever expected to include

ycanardeau commented on VocaDb.ResXFileCodeGenerator/GlobalOptions.cs:
```diff
@@ -0,0 +1,89 @@
+
+namespace VocaDb.ResXFileCodeGenerator;
+
+public sealed class GlobalOptions
```
and this

davhdavh committed:
Use TextReader instead of StringReader, just in case we want to read from a file
Add micro optimization in check for a file being the central resx file

davhdavh commented:
It can now generate almost identical code to my 15 year old resx generator plugin, that I had to fix every single visual studio release 👎 
Just need it to support 
* Custom class names (postfix) (Also actually need this one for razor pages)
* Build speed still sucks even after updating to IIncrementalGenerator... It seems it doesn't actually use that to build things faster. I created [https://github.com/dotnet/roslyn/issues/59174](https://github.com/dotnet/roslyn/issues/59174) for this issue.

davhdavh commented:
Btw, is there any particular reason why you used `TargetPath` instead of the `Link` that is used by resource generation that comes with msbuild?

davhdavh committed:
Fix mess with instance located in wrong class and test cases for same

ycanardeau commented:
> Build speed still sucks even after updating to IIncrementalGenerator...

Would it be related to another part of the code rather than roslyn itself? I'm not so confident with my code when it comes to performance...

> Btw, is there any particular reason why you used TargetPath instead of the Link that is used by resource generation that comes with msbuild?

The `TargetPath` option is added by another person in #4, and I don't know the reason why.

davhdavh commented:
No, I benchmarked it... Just takes ~200msec to generate all 1124 files. It does however take ~13 sec to generate all of the resources files and ~13sec to compile all of the generated cs files and ~3sec to build assembly and copy satellite around for a total of about 30 sec (which is why I moved them to a separate assembly a decade ago, saving those 30 sec every single time I had to rebuild the project has saved me many many hours).

davhdavh committed:
Add classname postfix, respect Link which is used to generate embeddedfilename

davhdavh committed:
Ups, wrong version tag

davhdavh committed:
1. Add code gen for lookup instead of using System.Resources.ResourceManager (set version to 3.1)
2. Setup debugging support
3. Setup integration test support
4. Use new namespace style decls in generated code
5. Split stringbuildergenerator class into multiple files
6. Test code now uses same method of determining source as generator code

ycanardeau commented:
@davhdavh This PR is quite big and it will take some time to review it. (I'm glad, and I appreciate you that you've converted this source generator to IIncrementalGenerator.) Do you think it's possible to split it into multiple parts? I think it's best to merge this PR before starting working on version 3.1. By the way, do we need to bump the version number to 3.1 even though we haven't published the nuget package yet?

ycanardeau commented on VocaDb.ResXFileCodeGenerator/VocaDb.ResXFileCodeGenerator.csproj:
```diff
@@ -1,8 +1,8 @@
   <PropertyGroup>
     <TargetFramework>netstandard2.0</TargetFramework>
-    <LangVersion>preview</LangVersion>
+    <LangVersion>latest</LangVersion>
```
Is there any particular reason why you changed `LangVersion` to `latest`?

davhdavh replied:
Not really. The latest vs17.5 gave warnings about using preview of .net. And I changed it. Besides using preview versions for something that is used in production isn't a particularly good idea, even if it has no consequences

ycanardeau commented on VocaDb.ResXFileCodeGenerator/AnalyzerReleases.Unshipped.md:
```diff
@@ -0,0 +1,9 @@
+; Unshipped analyzer release
+; https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.CodeAnalysis.Analyzers/ReleaseTrackingAnalyzers.Help.md
+
+### New Rules
+Rule ID | Category | Severity | Notes
+--------|----------|----------|-------
+VocaDbResXFileCodeGenerator001 | ResXFileCodeGenerator | Warning | StringBuilderGenerator
+VocaDbResXFileCodeGenerator002 | ResXFileCodeGenerator | Warning | StringBuilderGenerator
+VocaDbResXFileCodeGenerator003 | ResXFileCodeGenerator | Error | StringBuilderGenerator
```
I don't quite understand the purpose of these two files `AnalyzerReleases.Unshipped.md` and `AnalyzerReleases.Shipped.md`. Should I move these rules (adding `## Release 3.1` before `### New Rules`) to `AnalyzerReleases.Shipped.md` *after* publishing a nuget package?

davhdavh replied:
No. I just used the built-in quick fix which made those files. Don't have a clear idea what they actually do.

DanielBaumert replied:
"Don't have a clear idea what they actually do." -  Me too.

What I found is https://github.com/dotnet/roslyn-analyzers/blob/main/src/Microsoft.CodeAnalysis.Analyzers/ReleaseTrackingAnalyzers.Help.md but is it helpful? No

ycanardeau commented on README.md:
```diff
@@ -115,28 +135,341 @@ or globally
+Benefits of using `VocaDB` code generation:
+
+* All languages are placed in the main dll, no more satellite dlls
+* Lookup speed is ~600% faster (5ns vs 33ns)
```
Would you mind if I ask how to benchmark it? (I guess `launchSettings.json` is related to this?) I'm interested in it.

davhdavh replied:
I used benchmark.net and measured how long it took to do the 2 integration tests

ycanardeau commented on VocaDb.ResXFileCodeGenerator.sln.DotSettings:
```diff
@@ -0,0 +1,196 @@
+﻿<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
```
Personally, I haven't used ReSharper, so I'm wondering what the purpose of this file is. Is this different from `.editorconfig`?

davhdavh replied:
No. Pretty much the same. Although i did autogen the current .editorconfig from resharper, so there shouldn't be any settings that differ. I do highly recommend resharper, it has a ROI measured in days.

ycanardeau reviewed (approved):
Looks good. Is this PR still a work in progress?

ycanardeau commented on VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs:
```diff
@@ -1,140 +1,207 @@
 
-		builder.Append("    ");
 		builder.Append(options.PublicClass ? "public" : "internal");
+		builder.Append(options.PartialClass ? " partial" : "");
 		builder.Append(options.StaticClass ? " static class " : " class ");
```
The `partial` modifier can only appear immediately before `class` ([CS0267](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0267)).

davhdavh replied:
Will fix

ycanardeau commented on VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs:
```diff
@@ -1,140 +1,207 @@
+			builder.Append(options.InnerClassVisibility == InnerClassVisibility.SameAsOuter
+				? options.PublicClass ? "public" : "internal"
+				: options.InnerClassVisibility.ToString().ToLowerInvariant());
+			builder.Append(options.PartialClass ? " partial" : "");
+			builder.Append(options.StaticClass ? " static class " : " class ");
```
here too

davhdavh commented:
I have a small outstanding issue with what to do about dataannotations when not generating static members. But that isn't an issue that should prevent release. And I haven't found a solution yet.

With regards to version numbers, for a project like this it is more about optics than anything else. People might go and read the release notes on a full version than a simple 0.0.1 version bump and see the nice new features.

davhdavh committed:
Put partial the right position

davhdavh committed:
Broken unittest

ycanardeau commented:
Thanks a lot for the PR!
